src/rate_limiter.rs

```diff
@@ -0,0 +1,141 @@
+// KIND, either express or implied.
+//
+// Please review the Licences for the specific language governing permissions and limitations
+// relating to use of the SAFE Network Software.
+
+#[cfg(feature="fake_clock")]
+use fake_clock::FakeClock as Instant;
+use messages::{Request, UserMessage, UserMessageCache};
+use sha3::Digest256;
+use std::cmp;
+use std::collections::HashMap;
+use std::net::IpAddr;
+use std::time::Duration;
+#[cfg(not(feature="fake_clock"))]
+use std::time::Instant;
+
+/// Maximum total units the `RateLimiter` can hold at any given moment.
+const CAPACITY: u64 = 20 * 1024 * 1024;
+/// The number of units per second the `RateLimiter` will "leak".
+const RATE: f64 = 20.0 * 1024.0 * 1024.0;
+/// Default charge for a client get request.
+const DEFAULT_CLIENT_GET_CHARGE: u64 = 2 * 1024 * 1024;
+
+pub struct RateLimiter {
+    used: HashMap<IpAddr, u64>,
+    last_checked: Instant,
+}
+
+impl RateLimiter {
+    pub fn new() -> Self {
+        RateLimiter {
+            used: HashMap::new(),
+            last_checked: Instant::now(),
+        }
+    }
+
+    pub fn add_message(&mut self,
+                       online_clients: u64,
+                       ip_addr: IpAddr,
+                       hash: &Digest256,
+                       part_count: &u32,
+                       part_index: &u32,
+                       payload: &[u8])
+                       -> Result<(), ()> {
+        self.purge();
+        let total_used: u64 = self.used.values().sum();
+        let used = self.used.entry(ip_addr).or_insert(0);
+        let allowance = cmp::min(CAPACITY - total_used, CAPACITY / online_clients - *used);
+
+        let mut user_msg_cache = UserMessageCache::with_expiry_duration(Duration::from_secs(10));
+        let delta = if let Some(user_message) =
+            user_msg_cache.add(*hash, *part_count, *part_index, payload.to_vec().clone()) {
+            match user_message {
+                UserMessage::Request(request) => {
+                    match request {
+                        Request::GetIData { .. } |
+                        Request::GetMData { .. } |
+                        Request::GetMDataVersion { .. } |
+                        Request::GetMDataShell { .. } |
+                        Request::ListMDataKeys { .. } |
+                        Request::ListMDataValues { .. } |
+                        Request::ListMDataEntries { .. } |
+                        Request::GetMDataValue { .. } |
+                        Request::ListMDataPermissions { .. } |
+                        Request::ListMDataUserPermissions { .. } |
+                        Request::ListAuthKeysAndVersion { .. } |
+                        Request::GetAccountInfo { .. } |
+                        Request::Refresh(..) => DEFAULT_CLIENT_GET_CHARGE,
+                        _ => payload.len() as u64,
+                    }
+                }
+                _ => payload.len() as u64,
```

Comment:
This is not an allowed RPC for clients - probably need to disconnect from it ? Am wondering if all such clients should be kept in an LRUSet<IP> and not allowed to reconnect for a long time.